### ISSUE TITLE
Feature Request: fast mode that doesn't require compilation

### ISSUE THREAD

**Issue by ntc2** [haskell/haddock#612]
I would like a "fast" mode for Haddock. Often when I'm writing Haddocks, I just want to quickly build the docs and see how they render, and find out if I've introduced any parse errors (e.g. `--^` on GADT constructors: https://github.com/haskell/haddock/issues/43). In this use case I don't care about linked references, for example. I can save that for a Hackage upload or other "final form".

I don't know how Haddock works internally, but I see that right now it compiles my entire program every time I generate the docs. This takes several minutes for a large project. My hope is that by eliminating compilation, at the cost of linked references and perhaps other things, the documentation generation process could be greatly sped up (naively, the doc generation time should be similar to the parse time in this case).

If the Haddock maintainers are potentially open to such a `haddock --fast` feature, I might be interested in developing it myself. I'll have to learn about Haddock internals/implementation before I know how large of an undertaking this is, but if you already know this is something you're against, then I won't bother. Please let me know.

I searched the existing issues and found some issues related to performance, but nothing specific to avoiding compilation: https://github.com/haskell/haddock/issues/484, https://github.com/haskell/haddock/issues/345.

**Comment by alexbiehl** [comment]
@ntc2 I like the idea of a faster turnaround times. But this is tricky. Haddock works like this:

- Load the interface files for all dependencies; called `InterfaceFile` (see `Haddock.InterfaceFile.InterfaceFile`)
- Parse, type check and rename each module we want docs for
 This is the most expensive part, type checking and renaming one module implies checking all its dependencies transitively. 
- Build an `Interface` value (see `Haddock.Types.Interface`) for each module
- Take all read `InterfaceFile`s and `Interface`s to the render step to render the docs

Maybe it is possible to parse only (skipping type checking and renaming) a module and output a "draft" on how documentation might look like. Overall I think we would need to discuss the details on this may work out. Maybe you can find a sweetspot in the design space.

**Comment by ntc2** [comment]
@alexbiehl, thanks for explaining how Haddock works, that's very helpful. I notice you mention type checking, but not compilation proper. So I guess type checking is the bottle neck, so I'll look into eliminating that.

Yes, the fast mode would only produce a "draft", since e.g. the hyperlinking probably wouldn't work if type checking were eliminated. Naively, in fast mode I want to reduce the compilation dependency to only parsing. I would think that would provide enough info to generate rendered documentation, but without hyperlinks.

This weekend I'll plan to start looking into modules you referenced above, and how I might eliminate the type checking.

**Comment by Fuuzetsu** [comment]
Problem is with renaming, you have to tyoecheck to rename and to typecheck we load module and its dependencies too. I spent some time there before and it's not really viable to remove; it'd be possible to parse but most of the docs would be missing such as links, instance info, I think even more. 

The best way to improve the situation is to not load twice. Currently Haddock is a separate step but all it does is is call GHC with few flags changed. Maybe if we can invert control somehow or get GHC to dump out all the info we need by default then we can do something. Notably if we can have a flag that says "ghc please dump out intermediate typechecked module result please" (basically serialise a structure we need + comment info) then I think we can perhaps work with that. A problem is that rename environment will be empty so might have to figure out if it could be populated after the fact. I'm sceptical to be honest...

**Comment by ntc2** [comment]
I got caught up on low-hanging fruit, i.e. improving the documentation: https://github.com/haskell/haddock/pull/614, so I won't be making any progress on a `haddock --fast` this weekend. Really convenient to edit docs when the doc compilation time is so fast ;)

@Fuuzetsu , let me emphasize that the `haddock --fast` I have in mind will only be a "rough draft" mode (so maybe `haddock --fast` is the wrong name). The use case I care about is when I'm editing Haddocks, making minor changes, and I get frustrated because the full documentation build takes minutes. In these cases I'm not worried about producing the final form documentation, just seeing how my markup renders (e.g. I regularly do things like write `/` in `@` code blocks and then get confused when I get italics), and if there are any parse errors (in my experience at work, people regularly write Haddocks but rarely run `haddock`, and so sometimes I get parse errors in other peoples Haddocks).

So, in my "fast" use case I don't really care if the renames are wrong or the re-exports are missing, I just need to see how the markup itself rendered. I would of course love it if a full `haddock` run could be incrementally fast, but I believe you guys when you say that would no small task. So, to start, I'm just going to look into doing the simplest thing that solves my problem.

**Comment by Fuuzetsu** [comment]
I see. You want some --validate thing with oitput. It should be possible to make it fast-ish. Unsure how much work it will be to get renderer to take such partial info but shouldn't be _too_ bad I think.

**Comment by alexbiehl** [comment]
@Fuuzetsu Right, there already is: https://ghc.haskell.org/trac/ghc/ticket/2168. Which suggests GHC to pick up haddock docs. Haddock then just reads the interface and does its duty. This would also allow ghci to give documentations for names.

**Comment by Fuuzetsu** [comment]
Parsing only is quick enough where that ticket is not a requirement tbh; most work is in loading the whole code which we won't need for this

**Comment by siddhanathan** [comment]
Is someone still working on this?

**Comment by ntc2** [comment]
@siddhanathan , I never made any real progress and have not looked at it since May.

**Comment by Fuuzetsu** [comment]
I don't think anyone even started.

**Comment by mightybyte** [comment]
I've recently been feeling the pain of haddock's slowness (also referenced in issue #484).  Of course it would be better if it was faster.  But I've been wondering about an alternative approach for addressing this problem.  Could the haddock build be done incrementally?

**Comment by Fuuzetsu** [comment]
There's on-going GSOC hoping to address speed problem as part of it: https://summerofcode.withgoogle.com/projects/#4975710121230336
